sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala

```diff
@@ -0,0 +1,199 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import scala.collection.Map
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.analysis._
+import org.apache.spark.sql.types._
+
+object GetField {
+  /**
+   * Returns the resolved `GetField`. It will return one kind of concrete `GetField`,
+   * depend on the type of `child` and `fieldExpr`.
+   */
+  def apply(
+      child: Expression,
+      fieldExpr: Expression,
+      resolver: Resolver): GetField = {
+
+    (child.dataType, fieldExpr) match {
+      case (StructType(fields), Literal(fieldName, StringType)) =>
+        val ordinal = findField(fields, fieldName.toString, resolver)
+        SimpleStructGetField(child, fields(ordinal), ordinal)
+      case (ArrayType(StructType(fields), containsNull), Literal(fieldName, StringType)) =>
+        val ordinal = findField(fields, fieldName.toString, resolver)
+        ArrayStructGetField(child, fields(ordinal), ordinal, containsNull)
+      case (_: ArrayType, _) if fieldExpr.dataType.isInstanceOf[IntegralType]  =>
+        ArrayOrdinalGetField(child, fieldExpr)
+      case (_: MapType, _) =>
+        MapOrdinalGetField(child, fieldExpr)
+      case (otherType, _) =>
+        throw new AnalysisException(
+          "GetField is not valid on child of type " +
```

Comment:
one way I can think of to make the error messages more friendly is to specialize it for each child.datatype.